kvscheduler/descriptor/adapter-generator/template.go

```diff
@@ -0,0 +1,365 @@
+)
+
+////////// type-safe key-value pair with metadata //////////
+
+type {{ .DescriptorName }}KVWithMetadata struct {
+	Key      string
+	Value    {{ .ValueT }}
+	Metadata {{ .MetadataT }}
+	Origin   ValueOrigin
+}
+
+////////// type-safe Descriptor interface //////////
+
+type {{ .DescriptorName }}DescriptorAPI interface {
+	GetName() string
+	KeySelector(key string) bool
+	NBKeyPrefixes() []string
+	WithMetadata() (withMeta bool, customMapFactory MetadataMapFactory)
+{{- if .IsProtoValue }}
+	Build(key string, valueData {{ .ValueDataT }}) (value ProtoValue, err error)
+{{- else }}
+	Build(key string, valueData {{ .ValueDataT }}) (value {{ .ValueT }}, err error)
+{{- end }}
+	Add(key string, value {{ .ValueT }}) (metadata {{ .MetadataT }}, err error)
+	Delete(key string, value {{ .ValueT }}, metadata {{ .MetadataT }}) error
+	Modify(key string, oldValue, newValue {{ .ValueT }}, oldMetadata {{ .MetadataT }}) (newMetadata {{ .MetadataT }}, err error)
+	ModifyHasToRecreate(key string, oldValue, newValue {{ .ValueT }}, metadata {{ .MetadataT }}) bool
+	Update(key string, value {{ .ValueT }}, metadata {{ .MetadataT }}) error
+	Dependencies(key string, value {{ .ValueT }}) []Dependency
+	DerivedValues(key string, value {{ .ValueT }}) []KeyValuePair
+	Dump(correlate []{{ .DescriptorName }}KVWithMetadata) ([]{{ .DescriptorName }}KVWithMetadata, error)
+	DumpDependencies() []string
+}
+
+////////// Descriptor base implementation //////////
+
+// {{ .DescriptorName }}DescriptorBase provides default(=empty) implementations
+// for all the methods to extend from.
+type {{ .DescriptorName }}DescriptorBase struct {
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) GetName() string {
+	return "base"
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) KeySelector(key string) bool {
+	return false
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) NBKeyPrefixes() []string {
+	return nil
+}
+
+func (db *{{ .DescriptorName }}DescriptorBase) WithMetadata() (withMeta bool, customMapFactory MetadataMapFactory) {
+	return false, nil
+}
+
+{{ if .IsProtoValue -}}
+func (db *{{ .DescriptorName }}DescriptorBase) Build(key string, valueData {{ .ValueDataT }}) (value ProtoValue, err error) {
+{{- if .FromDatasync }}
+	// You can override specific methods of ProtoValue using embedding:
+	//
+	//	type MyProtoValue struct {
+	//		ProtoValue
+	//      typedMsg {{ .ValueDataT }}
+	//	}
+	//	
+	//	// ... (override some methods)
+	//
+	//  return &MyProtoValue{ProtoValue: NewProtoValue(valueData), typedMsg: valueData}	, nil
+	return NewProtoValue(valueData), nil
+{{- else }}
+	return nil, nil
+{{- end }}
+}
+{{- else }}
+func (db *{{ .DescriptorName }}DescriptorBase) Build(key string, valueData {{ .ValueDataT }}) (value {{ .ValueT }}, err error) {
+	return nil, nil
+}
+{{- end }}
+
+func (db *{{ .DescriptorName }}DescriptorBase) Add(key string, value {{ .ValueT }}) (metadata {{ .MetadataT }}, err error) {
+	fmt.Printf("Create for key=%s is not implemented\n", key)
```

Comment:
Instead of printing `not implemented` message, we could return an error that it is not implemented, otherwise the callers of the method will assume it succeeded.

Reply:
Actually it is OK not to implement a descriptor method. The printf statements I used only for testing, so I guess they can be removed.